pkg/downloader/downloader.go

```diff
@@ -0,0 +1,49 @@
+package downloader
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/enterprise-contract/gogather/pkg/filesystem"
+	"github.com/enterprise-contract/gogather/pkg/git"
+	"github.com/enterprise-contract/gogather/pkg/http"
+	"github.com/enterprise-contract/gogather/pkg/oras"
+	"github.com/spf13/afero"
+)
+
+// Downloader is an interface that defines the Download method.
+type Downloader interface {
+	Download(ctx context.Context, source string, destination string) error
```

Comment:
In a subsequent revision we'd like to provide metadata for the clients:
```suggestion
	Download(ctx context.Context, source string, destination string) (Metadata, error)
```

Where Metadata could be as simple as `type Metadata map[string]string`